src/runtime/cnv.r

```diff
@@ -1114,11 +1121,65 @@ static int ston(dptr sptr, union numeric *result)
 #endif                                  /* COMPILER */
       if (!realflag) {
          int rv;
-         rv = bigradix((int)msign, 10, ssave, end_s, result);
-         if (rv == RunError)
-            fatalerr(0, NULL);
-         return rv;
+         /* There are two possible reasons for being here:
+          *    The number (without a suffix) is larger then MaxLong.
+          *    The number overflowed whilst multiplying by the suffix.
+          *
+          * bigradix() does not support suffixes, so we compute the number
+          * sans suffix and multiply it (using big arithmetic) afterwards.
+          */
+         rv = bigradix((int)msign, 10, ssave, (esave ? esave : end_s), result);
+         if (rv == RunError) fatalerr(0, NULL);
+
+         if (esave) { /* We have a suffix: use bigmul (da * db -> dx) */
+           tended struct descrip da = nulldesc, db = nulldesc, dx =nulldesc;
+
+           /* Initialise da from the result returned by bigradix */
+           if (rv == T_Lrgint) {
+             da.dword = D_Lrgint;
+             BlkLoc(da) = (union block *)(result->big);
+           } else {
+             da.dword = D_Integer;
+             IntVal(da) = result->integer;
+           }
+
+           /* Initialise db from the suffix
+            * 32 bit m/c  [KMG] are  small integers, [TP] are large integers
+            * 64 bit m/c  [KMGTP] are all small integers
+            */
+           switch (*esave) {
+           default: return CvtFail; /* unkown suffix */
+           case 'k': case 'K': { MakeInt(1024L,&db); break; }
+           case 'm': case 'M': { MakeInt(1024L*1024,&db); break; }
+           case 'g': case 'G': { MakeInt(1024L*1024*1024,&db); break; }
+#if WordBits == 64
+           case 't': case 'T': { MakeInt(1024L*1024*1024*1024,&db); break; }
```

Comment:
`L` should be dropped from all cases because it forces a `long int` evaluation which is not the same on all platforms.